File: src/main/java/io/github/alexiscomete/lapinousecond/useful/FullTransaction.java
package io.github.alexiscomete.lapinousecond.useful;

import io.github.alexiscomete.lapinousecond.Main;
import io.github.alexiscomete.lapinousecond.entity.Player;
import io.github.alexiscomete.lapinousecond.view.AnswerEnum;
import org.javacord.api.entity.channel.TextChannel;
import org.javacord.api.interaction.MessageComponentInteraction;

import java.util.function.Consumer;
import java.util.function.Supplier;

public class FullTransaction extends VerifTransaction {
    public FullTransaction(Consumer<Double> addMoney, Consumer<Double> removeMoney, Supplier<Double> getMoney, Player p) {
        super(addMoney, removeMoney, getMoney, p);
    }

    public void askQuantity(Consumer<Double> after, TextChannel textChannel) {
        textChannel.sendMessage(p.getAnswer(AnswerEnum.ASK_MONTANT, true));

    }

    public void askQuantity(Consumer<Double> after, MessageComponentInteraction messageComponentInteraction) {
        messageComponentInteraction.createImmediateResponder().setContent(p.getAnswer(AnswerEnum.ASK_MONTANT, true));
    }

    private void addL(TextChannel textChannel, Consumer<Double> after) {
        Main.getMessagesManager().addListener(textChannel, p.getId(), (messageCreateEvent) -> {
            try {
                double d = Double.parseDouble(messageCreateEvent.getMessage().getContent());
            }
        });
    }

    public void full(TextChannel textChannel) {

    }

    public void full(MessageComponentInteraction messageComponentInteraction) {

    }
}

File: src/main/java/io/github/alexiscomete/lapinousecond/useful/FullTransaction.kt
package io.github.alexiscomete.lapinousecond.useful

import io.github.alexiscomete.lapinousecond.entity.Player
import io.github.alexiscomete.lapinousecond.messagesManager
import io.github.alexiscomete.lapinousecond.view.AnswerEnum
import org.javacord.api.entity.channel.TextChannel
import org.javacord.api.event.message.MessageCreateEvent
import org.javacord.api.interaction.MessageComponentInteraction

open class FullTransaction(
    addMoney: (Double) -> Unit,
    removeMoney: (Double) -> Unit,
    getMoney: () -> Double,
    p: Player,
    private val max: () -> Double,
) : VerifTransaction(addMoney, removeMoney, getMoney, p) {

    private fun askQuantity(after: (Double) -> Unit, textChannel: TextChannel) {
        textChannel.sendMessage(p.getAnswer(AnswerEnum.ASK_MONTANT, true))
        addL(textChannel, after)
    }

    private fun askQuantity(after: (Double) -> Unit, messageComponentInteraction: MessageComponentInteraction) {
        messageComponentInteraction.createImmediateResponder().setContent(p.getAnswer(AnswerEnum.ASK_MONTANT, true))
            .respond()
        addL(messageComponentInteraction.channel.get(), after)
    }

    private fun addL(textChannel: TextChannel, after: (Double) -> Unit) {
        messagesManager.addListener(textChannel, p.id) { messageCreateEvent: MessageCreateEvent ->
            try {
                val d = messageCreateEvent.message.content.toDouble()
                if (d > max()) {
                    messageCreateEvent.message.reply(p.getAnswer(AnswerEnum.VALUE_TOO_HIGH, true, max))
                    addL(textChannel, after)
                }
                after(d)
            } catch (e: IllegalArgumentException) {
                messageCreateEvent.message.reply(p.getAnswer(AnswerEnum.FORM_INVALID, true))
                addL(textChannel, after)
            }
        }
    }

    fun full(textChannel: TextChannel) {
        askQuantity({ aDouble: Double? -> askVerif(aDouble!!, textChannel) }, textChannel)
    }

    fun full(messageComponentInteraction: MessageComponentInteraction) {
        askQuantity(
            { aDouble: Double? -> askVerif(aDouble!!, messageComponentInteraction.channel.get()) },
            messageComponentInteraction
        )
    }
}